[docs] group cli commands; add missing ones

tnguyen-rh opened the pull request:
- add command groups headings as well as linked (toc-ish) intro blurb
- use hash-at-bol instead of line of equal-signs/hyphens for heading markup
- use backtick to markup hyphen (to mean stdin)
- remove noisy ‘Examples’ headings
- remove back-to-back codeblock, moving first one inline into intro blurb
- zonk ‘oc namespace’ (per https://github.com/openshift/origin/issues/3165#issuecomment-114315963)
- (oc login): add description, example
- (oc logout): add description
- (oc whoami): add description, example, "see also"
- (oc volume): add description, example
- (oc env): add description, example
- (oc config): add "synonym redirect"
- (oc export): add description, options table, example
- (oc import-image): add description, example
- (oc deploy): add description, examples
- (oc rollback): add description, examples
- (oc new-build): add description, component table, options table, example
- (oc cancel-build): add description, example
- (oc tag): add description, example
- (oc edit): add description, options table, example
- (oc exec): add "synonym redirect"
- (oc port-forward): add "synonym redirect"
- (oc proxy): add "synonym redirect"
- (oc new-project): add description, example
- (oc start-build): add "see also"
- (oc new-app): add description, options table, example, "see also"
- (oc status): add description, "see also"
- (oc project): add description, example

kargakis commented on docs/cli.md:
```diff
@@ -29,8 +34,281 @@ Common Flags
+
+### oc new-app
+
+This is a superset of [`oc new-build`](#oc-new-build).
```
I am not sure we should present `new-app` as a superset of `new-build`. Have a look at the command [description](https://github.com/openshift/origin/blob/f21a80e484e2ccc705a74a815df4a48e536c1acc/pkg/cmd/cli/cmd/newapp.go#L32-L45) and reword this likewise or use that as is.

kargakis commented on docs/cli.md:
```diff
@@ -29,8 +34,281 @@ Common Flags
+
+### oc new-build
+
+This creates a new buildConfig, optionally named, that references a docker image, a source code repository, or both.
```
Use the command [description](https://github.com/openshift/origin/blob/f21a80e484e2ccc705a74a815df4a48e536c1acc/pkg/cmd/cli/cmd/newbuild.go#L18-L27), which has more info.

kargakis commented on docs/cli.md:
```diff
@@ -29,8 +34,281 @@ Common Flags
+
+### oc build-logs
+
+> **NOTE**: This command will be later replaced by upstream (see: [kubectl log](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/kubectl_logs.md#kubectl-logs)).
```
Please change kubectl log (deprecated) to kubectl logs. I had fixed that link before but forgot to change the actual command name.

liggitt replied:
build-logs are build oriented… not sure that will be replaced with the logs command...

kargakis replied:
If `kubectl logs` uses the resource builder then it will essentially become a superset of `build-logs`

tnguyen-rh replied:
@kargakis 
Thanks for pointing out this discrepency.
Now the link text reads `kubectl logs`.

kargakis commented on docs/cli.md:
```diff
@@ -80,186 +356,217 @@ $ oc get pods -s template=production
 
 This command will return only pods whose `labels` include `"template": "production"`
 
```
Prompt users here to see `oc describe` and do the same for oc describe to `oc get`.

kargakis commented on docs/cli.md:
```diff
@@ -80,186 +356,217 @@ $ oc get pods -s template=production
+### oc expose
 
-#### Examples
+Expose containers internally as services or externally via routes. There is also
```
Start this as every other command in this doc. "This command exposes containers..."

kargakis commented on docs/cli.md:
```diff
@@ -80,186 +356,217 @@ $ oc get pods -s template=production
+### oc port-forward
 
-#### Examples
+This is synonymous with the [`kubectl port-forward`](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/kubectl_port-forward.md) command.
```
get, create, describe and other commands are also coming straight from Kubernetes but we give a description about them instead of pointing to kubectl doc. So to be consistent we either have to redirect all wrapped commands to Kube docs or describe them. I am in favor of the later option. See the descriptions of every wrapped command in [here](https://github.com/openshift/origin/blob/master/pkg/cmd/cli/cmd/wrappers.go).

liggitt commented on docs/cli.md:
```diff
@@ -29,8 +34,281 @@ Common Flags
+
+### oc build-logs
+
+> **NOTE**: This command will be later replaced by upstream (see: [kubectl logs](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/kubectl_logs.md#kubectl-logs)).
```
I still don't think we should position this as a soon to be deprecated command.

tnguyen-rh commented:
@kargakis @liggitt 
PTAL.  I have addressed most the points raised.  Unaddressed are:
* kubectl wrappers still indirect (linking to upstream docs);
* the "deprecated" blurb is still there -- shall i remove it?

liggitt commented:
I don't see the deprecated blurb in the diff any more

kargakis commented:
Replace the kubectl pointers with each command's usage as discussed on IRC and this LGTM

kargakis commented on docs/cli.md:
````diff
@@ -29,17 +34,303 @@ Common Flags
+$ oc rollback deployment-1 --output=json | oc update dc deployment -f -
+```
+
+See also [`oc update`](#oc-update).
````
Actually you need to update this to `oc replace` (was renamed).

tnguyen-rh replied:
@kargakis Thanks for catching this.  See: https://github.com/tnguyen-rh/origin/commit/d184f4411efafe3363113da06e1415481da67fcf

tnguyen-rh commented:
@liggitt Yes, the deprecated blurb was removed in: https://github.com/tnguyen-rh/origin/commit/ee686f4dc3a64e1ed950160a9cf42da39d21284b

tnguyen-rh commented:
@kargakis Is https://github.com/openshift/origin/commit/d4dfadd0b9575beb7703eaeb4151f6adcaa757e4 what you have in mind?

tnguyen-rh commented:
@liggitt @kargakis PTAL
For oc {config, port-forward, exec, proxy}, i've replaced the "synonymous" blurb w/ descriptions (sometimes w/ options table), and examples.

soltysh commented on docs/cli.md:
```diff
@@ -78,188 +369,320 @@ by applying key-value pairs that will be matched with the resource labels:
+
+This starts an editor opened to the YAML representation of the specified object.
+When the editor exits, the object is updated.
+You can specify the editor through environment variables `OSC_EDITOR`, `GIT_EDITOR`, or `EDITOR`.
```
OC_EDITOR

tnguyen-rh replied:
@soltysh Thanks for catching this.  Fixed in: https://github.com/tnguyen-rh/origin/commit/69679ce4ded53f5b690ff033043c04818b09a4ad

soltysh commented on docs/cli.md:
```diff
@@ -29,17 +34,303 @@ Common Flags
+
+### oc new-project
+
+This creates a new project, with you as the project admin.
```
s/with you/with current user
or eventually with currently logged in user...

kargakis replied:
This needs to get updated first in the new-project help message and then back here. I am ok with it for now.

tnguyen-rh replied:
@soltysh @kargakis I've gone ahead: https://github.com/tnguyen-rh/origin/commit/e34c1cab6bfd0d9b62140935755cd291bbb31be1 (but can revert if necessary).

soltysh commented on docs/cli.md:
````diff
@@ -29,17 +34,303 @@ Common Flags
+The general form is:
+
+```
+$ oc new-app <component> [options]
````
I'm missing here broad discussion what component might look like:
* just image, image stream
* just repository
* image~repository

etc.

kargakis replied:
new-app (IMAGE | IMAGESTREAM | TEMPLATE | PATH | URL ...)

tnguyen-rh replied:
`oc new-build` has a table.  How about we:
* create section "components" early on in the document
* mention there that `oc new-app` and `oc new-build` (and ...) use such `<component>`s
* add xrefs to `oc new-app` and `oc new-build`

tnguyen-rh replied:
Another (simpler) idea is to xref from `oc new-app` to `oc new-build`, which is what i went w/ in the end.

kargakis commented:
A new command was recently added, [oc patch](https://github.com/openshift/origin/blob/9ff47c565130355cd8542a07f845526acad84192/pkg/cmd/cli/cmd/wrappers.go#L79-L86). Add that under advanced commands.

soltysh commented on docs/cli.md:
```diff
@@ -29,17 +34,303 @@ Common Flags
+|`--output` (`-o`) *format*    | *format* is one of: `json`, `yaml`, `template`, `templatefile` |
+|`--output-version` *version*  | Output with *version* (default api-version)        |
+|`--no-headers`                | For default output, don't print headers            |
+|`--output-template` *s*       | Template string (`-o template`) or path (`-o templatefile`) |
```
Maybe sort this by option name, this way it's easier to find stuff.

tnguyen-rh replied:
@soltysh Sure, np: https://github.com/tnguyen-rh/origin/commit/c5776987cef46280b9ca102665a005fb90ed1fc3

soltysh commented on docs/cli.md:
````diff
@@ -29,17 +34,303 @@ Common Flags
+
+```
+$ oc start-build ruby-sample-build
+$ oc start-build --from-build=ruby-sample-build-275d3373-c252-11e4-bc79-080027c5bfa9
````
Only the builds created from builds already have those nasty long names. The initial ones created from buildconfig will have just following sequence numbers, please use those in the examples in here. Much more readable imho, this might scare people out :wink:

tnguyen-rh replied:
@soltysh Agreed.  I chose "1" as the sequence number: https://github.com/tnguyen-rh/origin/commit/dd83542b29ec79be12182309e3fc70bbb26e2113

soltysh commented on docs/cli.md:
```diff
@@ -29,17 +34,303 @@ Common Flags
+
+Flag `--latest` starts a deployment.
+Flag `--retry` retries the latest failed deployment.
+Flag `--cancel` cancels the in-progress deployment.
```
Previously options where in the table, so I'd suggest having them in table here as well.

tnguyen-rh replied:
@soltysh Generally, when there are few options (say, less than 5), i just mention them in a short paragraph, so sometimes there are tables and sometimes there are short paragraphs.  However, in this case, i have done as you suggested: https://github.com/tnguyen-rh/origin/commit/a7b6a2e7c03bf2777d15a1cd2230d2fd0c2ec968

soltysh commented on docs/cli.md:
```diff
@@ -29,17 +34,303 @@ Common Flags
+Flag `--latest` starts a deployment.
+Flag `--retry` retries the latest failed deployment.
+Flag `--cancel` cancels the in-progress deployment.
+The following example shows how to cancel the deployment mentioned in the preceding example:
```
Don't assume people reading this up to bottom, explain it, like one is just reading this part.

tnguyen-rh replied:
@soltysh Good point.  Addressed as part of: https://github.com/tnguyen-rh/origin/commit/a7b6a2e7c03bf2777d15a1cd2230d2fd0c2ec968

soltysh commented on docs/cli.md:
```diff
@@ -29,17 +34,303 @@ Common Flags
+
+This reverts the pod and container configuration back to a previous deployment.
+Scaling and trigger settings are normally left as-is.
+Flag `--dry-run` displays what the rollback *would do* but does not perform the rollback.
```
Again flags in-line, where previously in table.

tnguyen-rh replied:
@soltysh PTAL: https://github.com/tnguyen-rh/origin/commit/7d33f08db292b5abf2448a573ae440137b7b771b

soltysh commented on docs/cli.md:
```diff
@@ -29,17 +34,303 @@ Common Flags
+* `templatefile`
+
+With `--output template`, option `-t <template-string>` specifies a template string.
+With `--output templatefile`, option `-t <filename>` specifies a template file.
```
templatefile is Go template, iirc, maybe copy here the link to theirs doc, like you did a couple of lines before.

soltysh commented on docs/cli.md:
```diff
@@ -29,17 +34,303 @@ Common Flags
+| *imagestream*   | Use the latest image in *imagestream*. |
+| *path*          | Use source code found at *path*. |
+| *url*           | Use source code found at *url*. |
+| *image*`~`*url* | Note the tilde `~` between *image* and *url*.  In this case the component is actually made of two sub-components, the *image* and the source code found at *url*.  Use the image as the base and arrange to build the source code on it. |
```
I'm missing that exact table in `new-app` as well.

soltysh replied:
Don't mix and match regular text with code fragments, so instead of `*image*'~'*url*` just do `*image~url*`

soltysh commented on docs/cli.md:
```diff
@@ -29,17 +34,303 @@ Common Flags
+
+This cancels a pending or running build.
+Flag `--dump-logs` means to display the build logs for the cancelled build.
+Flag `--restart` means to create a new build after the current build is cancelled.
```
Flags in table.

kargakis commented:
Needs rebase

soltysh commented on docs/cli.md:
```diff
@@ -29,17 +34,310 @@ Common Flags
 | --insecure-skip-tls-verify | Skip SSL certificate validation (will make HTTPS insecure) |
 | --help (-h)                | Display help for specified command |
```
Put the flags into backticks, so that they are formatted as code. See like we did that in [s2i](https://github.com/openshift/source-to-image/blob/master/docs/cli.md).

soltysh commented on docs/cli.md:
````diff
@@ -29,17 +34,310 @@ Common Flags
+$ oc deploy <deployment-config> [options]
+```
+
+If invoked without options, display the latest deployment for the specified *deployment-config*.
````
s/display/it displays

soltysh commented:
Rebase, a couple of small nits, squash and you'll be ready to go.

kargakis commented on docs/cli.md:
```diff
@@ -1,18 +1,23 @@
-oc command line interface
-==============================
+# oc command line interface
```
Make this **OpenShift Command Line Interface**

tnguyen-rh commented:
@kargakis @soltysh @liggitt PTAL
I think i've addressed all the feedback (thanks, btw).
Rebased, but not yet fully squashed.  Will squash after it passes review.

kargakis commented on docs/cli.md:
````diff
@@ -1,51 +1,361 @@
+
+For example:
+
+```
````
Add *bash* here next to the backticks like in L55 for the love of markdown

kargakis commented on docs/cli.md:
````diff
@@ -1,51 +1,361 @@
+The *template-string* is in [golang template](http://golang.org/pkg/text/template) format.
+Here are two examples:
+
+```
````
Add bash here too

kargakis commented on docs/cli.md:
````diff
@@ -84,188 +394,320 @@ by applying key-value pairs that will be matched with the resource labels:
+
+Some examples are:
 
 ```
````
Add bash here too

kargakis commented on docs/cli.md:
````diff
@@ -84,188 +394,320 @@ by applying key-value pairs that will be matched with the resource labels:
-$ oc expose dc ruby-hello-world --port=8080 --name=myservice --generator=service/v1
+Some examples are:
+
+```
````
Add bash here too

tnguyen-rh commented:
@kargakis OK (all annotated).  Anything else?

tnguyen-rh commented:
@kargakis Oops, almost forgot `oc patch`.

soltysh commented:
LGTM, please squash your commits into one and it can be merged. Eventual fixes can be addressed as a followups. Thanks @tnguyen-rh for doing this!

kargakis commented:
LGTM

Squash and it's ready to go

tnguyen-rh commented:
@kargakis @soltysh @liggitt 
Thanks everyone for the patient reviews.
Squashed!

soltysh commented:
[merge]

kargakis commented:
[merge] again
